homeassistant/components/habitica/switch.py

```diff
@@ -0,0 +1,138 @@
+    is_on_fn: Callable[[HabiticaData], bool]
+
+
+class HabiticaSwitchEntity(StrEnum):
+    """Habitica switch entities."""
+
+    SLEEP = "sleep"
+
+
+SWTICH_DESCRIPTIONS: tuple[HabiticaSwitchEntityDescription, ...] = (
+    HabiticaSwitchEntityDescription(
+        key=HabiticaSwitchEntity.SLEEP,
+        translation_key=HabiticaSwitchEntity.SLEEP,
+        device_class=SwitchDeviceClass.SWITCH,
+        turn_on_fn=lambda coordinator: coordinator.api["user"]["sleep"].post(),
+        turn_off_fn=lambda coordinator: coordinator.api["user"]["sleep"].post(),
+        is_on_fn=lambda data: data.user["preferences"]["sleep"],
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: HabiticaConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up switches from a config entry."""
+
+    coordinator = entry.runtime_data
+
+    async_add_entities(
+        HabiticaSwitch(coordinator, description) for description in SWTICH_DESCRIPTIONS
+    )
+
+
+class HabiticaSwitch(CoordinatorEntity[HabiticaDataUpdateCoordinator], SwitchEntity):
+    """Representation of a Habitica Switch."""
+
+    _attr_has_entity_name = True
+    entity_description: HabiticaSwitchEntityDescription
+
+    def __init__(
+        self,
+        coordinator: HabiticaDataUpdateCoordinator,
+        entity_description: HabiticaSwitchEntityDescription,
+    ) -> None:
+        """Initialize a Habitica switch."""
+        super().__init__(coordinator)
+        if TYPE_CHECKING:
+            assert coordinator.config_entry.unique_id
+        self.entity_description = entity_description
+        self._attr_unique_id = (
+            f"{coordinator.config_entry.unique_id}_{entity_description.key}"
+        )
+        self._attr_device_info = DeviceInfo(
+            entry_type=DeviceEntryType.SERVICE,
+            manufacturer=MANUFACTURER,
+            model=NAME,
+            name=coordinator.config_entry.data[CONF_NAME],
+            configuration_url=coordinator.config_entry.data[CONF_URL],
+            identifiers={(DOMAIN, coordinator.config_entry.unique_id)},
+        )
+
+    @property
+    def is_on(self) -> bool | None:
+        """Return the state of the device."""
+        return self.entity_description.is_on_fn(
+            self.coordinator.data,
+        )
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Turn the entity on."""
+        try:
+            await self.entity_description.turn_on_fn(self.coordinator)
+        except ClientResponseError as e:
+            if e.status == HTTPStatus.TOO_MANY_REQUESTS:
+                raise ServiceValidationError(
+                    translation_domain=DOMAIN,
+                    translation_key="setup_rate_limit_exception",
+                ) from e
+            raise HomeAssistantError(
+                translation_domain=DOMAIN,
+                translation_key="service_call_exception",
+            ) from e
```

Comment:
Instead of having this everywhere, maybe we can wrap the function. I think I implemented something similar in `linear_garage_door`